streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -0,0 +1,229 @@
+                                                                                  keySchema.upperRange(binarySessionId, latestSessionStartTime).get());
+        final KeyValueIterator<Bytes, byte[]> storeIterator = bytesStore.fetch(binarySessionId, earliestSessionEndTime, latestSessionStartTime);
+        final HasNextCondition<Bytes> hasNextCondition = keySchema.hasNextCondition(binarySessionId,
+                                                                                    earliestSessionEndTime,
+                                                                                    latestSessionStartTime);
+        final PeekingKeyValueIterator<Bytes, LRUCacheEntry> filteredCacheIterator = new FilteredCacheIterator(cacheIterator, hasNextCondition);
+        return new MergedSortedCacheKeyValueStoreIterator<>(filteredCacheIterator, storeIterator, serdes);
+    }
+
+
+    public void remove(final Windowed<K> sessionKey) {
+        put(sessionKey, null);
+    }
+
+    public void put(final Windowed<K> key, AGG value) {
+        final Bytes binaryKey = SessionKeyBinaryConverter.toBinary(key, keySerde.serializer());
+        final LRUCacheEntry entry = new LRUCacheEntry(serdes.rawValue(value), true, context.offset(),
+                                                      key.window().end(), context.partition(), context.topic());
+        cache.put(name, binaryKey.get(), entry);
+    }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, AGG> fetch(final K key) {
+        return findSessionsToMerge(key, 0, Long.MAX_VALUE);
+    }
+
+
+    public String name() {
+        return bytesStore.name();
+    }
+
+    @SuppressWarnings("unchecked")
+    public void init(final ProcessorContext context, final StateStore root) {
+        bytesStore.init(context, root);
+        initInternal((InternalProcessorContext) context);
+    }
+
+    @SuppressWarnings("unchecked")
+    void initInternal(final InternalProcessorContext context) {
+        this.context = context;
+
+        if (keySerde == null) {
+            keySerde = (Serde<K>) context.keySerde();
+        }
+
+
+        this.serdes = (StateSerdes<Windowed<K>, AGG>) new StateSerdes<>(bytesStore.name(),
+                                                                              new SessionKeySerde<>(keySerde),
+                                                                              aggSerde == null ? context.valueSerde() : aggSerde);
+
+
+        this.name = context.taskId() + "-" + bytesStore.name();
+        this.cache = this.context.getCache();
+        cache.addDirtyEntryFlushListener(name, new ThreadCache.DirtyEntryFlushListener() {
+            @Override
+            public void apply(final List<ThreadCache.DirtyEntry> entries) {
+                for (ThreadCache.DirtyEntry entry : entries) {
+                    final Bytes binaryKey = entry.key();
+                    final RecordContext current = context.recordContext();
```

Comment:
In previous stores, like `CachingKeyValueStore`, you put the code below in a separate method called `putAndMaybeForward` that I thought was a neat way to break the code into chunks.